go/sigid/sigid.go

```diff
@@ -0,0 +1,125 @@
+package sigid
+
+import (
+	"encoding/binary"
+	"sort"
+
+	"github.com/blang/semver"
+	"github.com/keybase/client/go/kbcrypto"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+type YNM int
+
+const (
+	No    YNM = 0
+	Yes   YNM = 1
+	Maybe YNM = 2
+)
+
+func isModernSigIDMaker(clientName string, clientVersionString string) YNM {
+	var clientVersion, cutoffVersion semver.Version
+	var err error
+	if clientName != "keybase.io go client" {
+		return Yes
+	}
+	clientVersion, err = semver.Make(clientVersionString)
+	if err != nil {
+		return Yes
+	}
+	cutoffVersion, err = semver.Make("1.0.16")
+	if err != nil {
+		panic(err)
+	}
+	if clientVersion.GT(cutoffVersion) {
+		return Yes
+	}
+	if clientVersion.EQ(cutoffVersion) {
+		return Maybe
+	}
+	return No
+}
+
+// isMaybeModernSigIDMakerModern is something. The general lay of the
+// land is that SigIDs generated with Keybase prior to version 1.0.16
+// did so incorrectly, since they didn't compute the SHA256 of the NaclSigInfo
+// properly before computing the SigID. Those versions of Keybase at 1.0.17
+// and after do the right thing. The problem is 1.0.16. Some clients with the
+// 1.0.16 compute SigIDs the right way, and some do it the wrong way,
+// (likely because Linux nightlies were going out with the 1.0.16 version).
+//
+// There are 77331 total signatures that were generated via 1.0.16, of which
+// 16555 where generated the wrong way (using the <= 1.0.16 method). In data.go,
+// we have prefixes of wrong/legacy signature IDs, that need to be fixed. To save
+// binary space, we only include in data.go the shortest possible prefix of the
+// wrong/legacy signature IDs that don't collide with prefixes from the group of
+// correct signatures. So the idea is that if we compute a hash the right way,
+// then find its prefix in data.go, then we need to recompute it the wrong way,
```

Comment:
confused here. Does "Right" mean modern and "Wrong" mean legacy (due to the `hash` bug)? There's no security issue with the "wrong" one right?

So does `data.go` contain what's in the merkle tree/db? Or is it the modern version for all of them?

Reply:
modern version of them.